### PR TITLE
Bug 1808358 - Refactor Onboarding Page to make it reusable

### PR DESCRIPTION
**Pull request by rahulsainani**
### What
– Create `OnboardingPage` that takes in `OnboardingPageUiState` to display the content. 
– Create `OnboardingContent` composable that selects the click actions and content based on the state and passes on the right content to OnboardingPage.

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation

Used by GitHub Actions.

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt:118]:
```suggestion
                recordImpressionEvent = {},
```

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:50]: Any thoughts on combining `OnboardingPageUiState` and `UpgradeOnboardingState` or making `UpgradeOnboardingState` a child class? It could simplify the `OnboardingPage(...)` initialization with the giant `when`.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:98]: do the metrics pings compile with the Compose previews?

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (MozillaNoah) [inline app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:64]: Curious what you think, @gabrielluong, but this seems like a Composable that's serving UI for 2+ screens because we don't have Compose-based navigation. The modularity is appreciated and the API seems straightforward. It's just not 100% clear to me if all onboarding screens are/will/should conform to this template.

**Review by MozillaNoah — COMMENTED** [review]
quick pass as I'm multitasking

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:50]: My understanding was that this work was motivated by https://www.figma.com/file/d7l1iGabQYqnLcnQbHDGzM/Wallpaper-2.0-%2B-Onboarding-v106?node-id=8924%3A220946&t=ihw1RVyD5zhJ7Gze-4 which is to add an onboarding screen for push notifications. Therefore, keeping `OnboardingPageUiState` might still make sense.

Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:50]: I have also briefly discussed with @rahulsainani about the pros and cons with the status quo with managing this `enum` state vs adopting a `NavHost` to get rid of the `when`. At the end of the day, both approaches would be  fine. I have filed https://github.com/mozilla-mobile/fenix/issues/28330 last week with the intent to move this to use navigation to remove the `when` that is going on.

Rahul have said "there's not a lot of navigation happening and we'd have to clear the back stack anyway as we don't allow users to go back on Welcome screen. It's fine either way imo". I am mostly okay with keeping the status quo since the navigation is minimal and not having to handle the back stack. We can keep https://github.com/mozilla-mobile/fenix/issues/28330 around for future exploration if we felt strongly about it.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:207]:
```suggestion
private fun UpgradeOnboardingScreenPreview() {
```
This is just to stick with the Compose Preview naming convention.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt:54]: I think we could remove comments like these. I would hope it's generally understood when someone sees `verticalArrangement = Arrangement.SpaceEvenly` rather than have to explain what a property of `verticalArrangement` is doing.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:98]: This is fair point.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:50]: Gotcha gotcha. I'm definitely open to there being a navigation refactor down the road, and I'm glad yall had that talk to figure out what makes sense. I'm open to moving forward as is, with the caution if onboarding, for some reason, gets interest in becoming more complex.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt:102]: Let's remove the comment here. I think we are repeating the code here with more or less the same exact words.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPageUiState.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPageUiState.kt:11]: Let's add the @param description.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:64]: I briefly eluded to this in https://github.com/mozilla-mobile/fenix/pull/28376/files#r1060997395, but I think the intent of this is to be used only in HomeOnboardingDialogFragment.kt.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:207]: I don't expect this to work nor do I think we need this.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:50]: @MozillaNoah were you suggesting something like this?

```
enum class OnboardingPageUiState(
    @DrawableRes val image: Int,
    @StringRes val title: Int,
    @StringRes val description: Int,
    @StringRes val primaryButtonText: Int,
    @StringRes val secondaryButtonText: Int? = null,
    val recordImpressionEvent: () -> Unit,
) {
    Welcome(
        image = ...
    ),
    SyncSignIn(
        image = ...
    )
}
```

If so, then I'd +1 it, as I think it well encapsulates the static differences between the pages well. I do think we will need to move the metrics callbacks into a a `Controller` class and passed to the Composables per gl's comment below in order to get previews rendering.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:98]: I am interested in continuing the discussion of simplifying the Controller/Interactor pattern, so I added it as a topic to our next experience team sync.

Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:98]: @MatthewTighe Sounds good to me. I think my ask would always be to start up a RFC - it doesn't have to be long or anything, but we just want to do some due diligence and provide a forum to address any concerns from the wider team.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPageUiState.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPageUiState.kt:13]: +1 on `onRecordImpression` convention. I do like the idea with following a callback naming convention that has been adopted in the wider project.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:50]: @MatthewTighe This is also an interesting idea.

Comment (rahulsainani) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:50]: The idea was that `OnboardingPage` is in a different component and doesn't need to who/what uses it, similar to a `Button`, the callers can keep the objects and pass on the data, as there might be more features using the OnboardingPage. Also makes it possible to move `OnboardingPage` to a different module, let me know your thoughts

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPageUiState.kt
Comment (rahulsainani) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPageUiState.kt:11]: Think the params are quite self descriptive :) Similar to the comment here --> https://github.com/mozilla-mobile/fenix/pull/28376#discussion_r1061006455

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (rahulsainani) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:207]: Hmm yeah, the idea was to start the function with Preview so it doesn't pollute autocomplete, let me know what you think? 
If there's a convention, can i read it somewhere?

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPageUiState.kt
Comment (rahulsainani) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPageUiState.kt]: I ended up adding the param docs as it might be used by other callers 👍

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (rahulsainani) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:207]: Btw the preview should work, I think maybe because metrics are now only called in side effects (click listeners and effect handlers), and they are not in the composition.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (rahulsainani) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:98]: Yeah the previews work, I think maybe because the metrics ping is in the side effects

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:207]: We have mostly adopted `Preview` being the suffix everywhere because that's how it is in the Compose Preview docs https://developer.android.com/jetpack/compose/tooling/studio#preview

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:50]: That is a good point. If we consider this with the possibility of modularizing Fenix in the future I think our options are:

1. Treat onboarding as a complete feature vertical, in which case my suggestion above would still work. Other features that required onboarding could come add their static properties in the onboarding module itself.
2. Define onboarding pages per feature module and pass them to the onboarding module, as it sounds like you're suggesting here. If we went this route, another convention you could consider is to define a top-level `val` for each onboarding page. For example, in `/feature/onboarding/Onboarding.kt` we could have `val welcomeOnboarding = OnboardingPageUiState(...)`.

I don't necessarily have a strong opinion here, but I would lean towards the 1st option still. My impression when onboarding is discussed here is that we will be treating it as a distinct feature in terms of additions, experimentation, etc. I think my example in 2 also illustrates what could become a slightly awkward module structure - it would make sense to have something like `/feature/sync/Onboarding` or `/feature/notifications/Onboarding`, but then those instances are split from any number of pages that could end up defined in `/feature/onboarding/Onboarding`.

It also really doesn't matter for now, I don't think, as this is all hypothetical work. Just wanted to share some of my thoughts. Plans for the future never survive first contact anyway, so I'd stick with whatever feels most comfortable to you 🙂

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:207]: Because the preview functions are private, they won't pollute the global namespace.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (MozillaNoah) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:50]: @MatthewTighe, yes exactly that!

Comment (rahulsainani) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:50]: @MatthewTighe thanks for the explanation and the thought process. I think that's a bit outside the scope of this PR. Also, Welcome and Sync Sign In Onboarding uses Indicators (which are not touched) and the upcoming notification pre permission dialog doesn't use it, that's why the smallest possible component (OnboardingPage) was extracted so we can reuse that part. I'm open to commonalising more parts of the UI but I would say we should do that going forward as things get more clear based on the usages. wdyt?

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt:88]: I think we can leave the named `id` parameter out of these `stringResource` calls

File: app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/onboarding/view/UpgradeOnboardingScreen.kt:64]: As gl mentioned in [this comment](https://github.com/mozilla-mobile/fenix/pull/28376/files#r1060997395), the pattern I usually see is to host top-level state and events in the parent fragment. This will make it easier to move these metrics calls into a controller as well. An example of that could be the [WallpaperSettingsFragment](https://searchfox.org/mozilla-mobile/rev/825ba5537cdc9901b5a62bd4e8f9016da29e7c9d/fenix/app/src/main/java/org/mozilla/fenix/settings/wallpaper/WallpaperSettingsFragment.kt#61).

Basically, I think we could remove the `UpgradeOnboardingScreen` entirely here, and call directly to `OnboardingContent` though it would likely make sense to change the name as well.

**Review by MatthewTighe — APPROVED** [review]
This is looking pretty good to me! I just had a couple more small things, including some discussion about moving some of these click handlers into a controller eventually. I think that could be done as a follow-up in order to get this moving again, but it would be good to file an issue for it. 

Since the time gap is so large, I'm going to approve for now to unblock you with the assumption that my few small concerns are addressed. That said, it might be worth running it by gl one more time if there is any overlap for the two of you.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt:88]: I have discussed this with @MozillaNoah a long time ago, but I think it's fine to just leave these `id` in since they are autofilled when you autocomplete `stringResource` and I rather not have to force developers to delete it.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPageState.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPageState.kt:18]:
```suggestion
 * @param title [StringRes] of the onboarding headline text.
 * @param description [StringRes] of the onboarding body text.
 * @param primaryButtonText [StringRes] of the primary button text.
 * @param secondaryButtonText [StringRes] of the secondary button text.
 * @param onRecordImpressionEvent callback for recording impression event.
```

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt:49]: **This is a blocking change**: This is unfortunately gonna change the layout from the designs in https://www.figma.com/file/d7l1iGabQYqnLcnQbHDGzM/Wallpaper-2.0-%2B-Onboarding-v106?node-id=6058%3A209483&t=bMa3YpVzO7xeRq9s-0. In particular, there's too much space between the Buttons and the Indicator when we would ideally want the Buttons to be at the bottom. We should ideally not change the layout from the desired specs in this refactoring.

Before (left) vs After (right):
Screen Shot 2023-01-05 at 10 10 02 PM

**Review by gabrielluong — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt:52]: I believe the solution to our image scale issue can be solved via `contentScale = ContentScale.FillBounds`. https://developer.android.com/reference/kotlin/androidx/compose/ui/layout/ContentScale.Companion#FillBounds()

We currently use this in https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/res/layout/fragment_turn_on_sync.xml#38.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt
Comment (rahulsainani) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt]: Good catch! removed it! The idea before the refactoring was to merge image and title with one parent so this might have made sense then, not anymore! 
Yeah this modifier merges the composables in the semantics tree, way better explanation than I can provide --> https://www.youtube.com/watch?v=iV0h_6G3CYk

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt
Comment (rahulsainani) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt:48]: Umm I think this specific preview size works best for the left option but a different preview size might work better for the option on the right. At this moment, none of these are ideal, but as we discussed in the [dimensions + box with constraints PR](https://github.com/mozilla-mobile/fenix/pull/28348) and slack, that we could use grids and ratios to solve the problem of supporting different screen sizes. Adding a hard coded dimension here to make the button move down would look good in our previews but won't on different screen sizes. If you have a concrete solution, I'm open.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt
Comment (rahulsainani) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt:51]: I'm afraid that doesn't work and it's not related to this change, let's move this thread to the relevant issue

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPageState.kt
Comment (rahulsainani) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPageState.kt:25]: We could've, it's just another approach, an advantage of using `Int`s is that these can be file level variables but if they're composables they can only be accessed from a composable scope.

File: app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt:48]: One approach would be to pursue the original `SpaceBetween` approach with the 3 containers being:
- Dismiss button
- Image and text
- Buttons and Indicator
I do understand this change will likely break some of the foundations how you have chosen to refactor each section.

The spacing between the Button and Indicator does fall between 24dp and 32dp depending on whether or not the Secondary Button is shown. Since 24 and 32 are actual spacing values that we can find in the [Spacing Spec](https://www.figma.com/file/pEyGeE4KV5ytYHeXMfLcEr/Mobile-Styles?node-id=1017%3A56444&t=MxgC2NnW8WhXmN0q-0), we should treat the spacing here as intentional. Using `SpaceEvenly` would not satisfy this requirement because the buttons would be anchored to the Column with image and text, and disconnected from the indicators. There's also 16dp spacer between the button and indicator in [OnboardingContent](https://github.com/mozilla-mobile/fenix/pull/28376/files#diff-cee11c3fc1e078c563b8bc30d76866b9f01ead3fe2f7e1deebf70e271744cf6dR161) so I would imagine the spacing between the button and indicator would never follow the spec even if we had a grid system. 

While I do understand what you're saying, we can't simply just deviate from the original design spec for this reasoning alone since we have already gotten prior UX review for this. 

My ask here is that we shouldn't change the layout from the status quo in a refactoring. If we want to address the layout and scaling issues, let's address that separately or fix it before this refactoring so we are better guided on how to best refactor this. The underlying problem that I still want to focus on is only the spacing between the indicators and buttons.

Comment (rahulsainani) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt]: Hmm, updated it, could you take a look again?

Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/onboarding/view/OnboardingPage.kt]: I discussed what I was still seeing with @rocketsroger and @MozillaNoah. There are still fundamental issues with how things are laid out that are breaking the design specs with the current approach. I would still suggest laying things out based on how I have described it and can be seen in the original implementation:

Column [Space Between]:
  - Dismiss button
  - Image and text
  - Buttons and Indicator

In your current implementation, we can't use `SpaceBetween` between the image and text because the spacing between them will now be dynamic. The image and text should be in the same container with a 32dp spacer between them to match the designs:
Screen Shot 2023-01-09 at 5 38 23 PM

The spacing between the button and indicators are also dependent on how many buttons are shown:
| One Button | Two Buttons |
|---|---|
|Screen Shot 2023-01-09 at 5 41 46 PM|Screen Shot 2023-01-09 at 5 41 51 PM|

The `Spacer` as the first item in this `SpaceBetween` Column is also a bit of a bandaid solution that is masking other problems, which I don't really recommend since the intent of the refactoring was to make all of this better. We have already drastically changed how items are wrapped in their various column containers and as a result the layout has also changed. We should proceed somewhat cautiously to avoid further review churn. I am aware this is high priority work, but my understanding from @rocketsroger was that there is still a path forward that does not depend on this refactoring.

**Review by gabrielluong — CHANGES_REQUESTED** [review]
See https://github.com/mozilla-mobile/fenix/pull/28376#discussion_r1065184222

**Comment by rahulsainani** [comment]
Closing this as #28630 and further PRs will address commonalising the `OnboardingPage` component.

**Pull request closed without merging**
